Implement BLL structure for Game entity

Oleksii-Stepanyk opened the pull request:
### This PR implements the general structure for BLL and Game entity-related classes/interfaces inside BLL.

### User stories:
   - Add Game logic for BLL

### Changes made:
   - Created generic `IService` and `IGameService` interface
   - Created `GameService` class without method implementation
   - Created `GameDTO` entity
   - Created `IMapper` interface and `GameMapper` class
   - Implemented `GameMapper` methods for mapping from/to `GameDTO` to/from `Game` entity 
   - Created `ServiceCollectionExtensions` class inside the BLL
   - Created `IDataGenerator` interface and `GameDataGenerator` for test data generation
   - Created `IRepositoryMockCreator` and `GameRepositoryMockCreator` classes for tests
   - Configured `GameService` unit tests environment

### Notes:
   1. Tests:
      - I decided to add method implementation in the future, along with the UI layer for each user story separately. For now, there is only a constructor that prepares a mock version of `GameRepository`.
Also, there is a `GameDataGenerator` that is used to generate test data for unit tests. This data is injected into `GameMockRepositoryCreator`, which configures a mock repository for unit tests. 
   2. Interfaces:
      - Since this is the first PR related to BLL, I created a generic IService interface, which can be used with any entity, and contains only basic CRUD operations. The same principle applies with the `IMapper` for mapping entities with DTOs and `IRepositoryMockCreator`  for test purposes.
   3. ServiceCollectionExtensions:
      - The same as for the DAL - this class will be used in the UI layer for DI. Instead of manually configuring all services and other things in one place, we can split them up by layer and call only one method per layer.

hermanhavva commented on OnlineGameStore.BLL/Interfaces/IGameService.cs:
```diff
@@ -0,0 +1,6 @@
+
+namespace OnlineGameStore.BLL.Interfaces;
+
+public interface IGameService : IService<Game, GameDto> { }
```
@OlenaKazymyrova 
Could you please leave some feedback on my comments so that we can cover those moments, and introduce comments if needed :)

OlenaKazymyrova replied:
Yep, you're absolutely right — great catch

Oleksii-Stepanyk replied:
That's the point, we need an entity object in repository methods, so I used mapping from DTO to the Entity directly.

OlenaKazymyrova replied:
yep, I get your point. We definitely need to use DAL entities inside the BLL — it's totally fine to map from DTOs to entities before passing them to the repository. The key thing is just to keep entities out of the BLL interfaces so other layers don’t depend on the DAL. 
I believe the full picture will be clearer once you implement the API layer — then you'll see how the different layers depend on each other

Oleksii-Stepanyk replied:
I only now understand what a horrible mistake I have made.
My implementation was expected to get the `Game` entity from DAL, but it is impossible to do this, because this method will be called from the UI layer.

Oleksii-Stepanyk commented on OnlineGameStore.BLL/DTOs/GameDto.cs:
```diff
@@ -0,0 +1,11 @@
+    public Guid? Publisher { get; set; }
+    public Guid? Genre { get; set; }
+    public Guid? License { get; set; }
+}
```
I named these fields as we discussed in our meetings.

Oleksii-Stepanyk commented on OnlineGameStore.BLL/Mappers/GameMapper.cs:
```diff
@@ -0,0 +1,44 @@
+    {
+        return source.Select(Map);
+    }
+}
```
Yep, that's a great point. Should we change it now or leave it for the next iteration?

OlenaKazymyrova replied:
I would say if we can complete everything planned for this sprint + AutoMapper, then let's include it in this iteration.
If not, let's create a US and implement it in the next iteration

Oleksii-Stepanyk commented:
### Update:
 - Added `AutoMapper` package and `MappingProfile` instead of custom `IMapper` and `GameMapper`
 - Changed `IService` and `GameService` implementation, it now expects DTO instead of Entity (That was my pure mistake)

hermanhavva commented on OnlineGameStore.BLL/Services/GameService.cs:
```diff
@@ -0,0 +1,42 @@
+    private readonly IGameRepository _repository;
+    private readonly IMapper _mapper;
+
+    public GameService(IGameRepository repository, IMapper mapper)
```
You pass mapper instance inside the constructor, but do not use it

Oleksii-Stepanyk replied:
Thank you for noticing!

hermanhavva reviewed (commented):
Good job! I would also suggest changing the naming in the game entity as well (see commnets)

Oleksii-Stepanyk commented on OnlineGameStore.BLL/Mapping/MappingProfile.cs:
```diff
@@ -0,0 +1,19 @@
+            .ForMember(dest => dest.GenreId, opt => opt.MapFrom(src => src.Genre))
+            .ForMember(dest => dest.LicenseId, opt => opt.MapFrom(src => src.License));
+    }
+}
```
Let's leave it for a future DAL improvement
